core/client/src/children.rs

```diff
@@ -0,0 +1,124 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+
+use std::collections::HashMap;
+use kvdb::{KeyValueDB, DBTransaction};
+use codec::{Encode, Decode};
+use crate::error;
+use std::hash::Hash;
+
+
+/// Map of children blocks stored in memory for fast access.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ChildrenMap<K, V> where 
+	K: Eq + Hash + Clone + Encode + Decode,
+	V: Eq + Hash + Clone + Encode + Decode,
+{
+	storage: HashMap<K, Vec<V>>,
+}
+
+impl<K, V> ChildrenMap<K, V> where
+	K: Eq + Hash + Clone + Encode + Decode,
+	V: Eq + Hash + Clone + Encode + Decode,
+{
+	/// Creates an empty children map.
+	pub fn new() -> Self {
+		Self {
+			storage: HashMap::new(),
+		}
+	}
+
+	/// Returns the hashes of the children blocks of the block with `parent_hash`.
+	pub fn hashes(&self, db: &KeyValueDB, column: Option<u32>, prefix: &[u8],
+		parent_hash: K) -> error::Result<Vec<V>> {
+
+		let mut buf = prefix.to_vec();
+		parent_hash.using_encoded(|s| buf.extend(s));
+
+		if let Ok(raw_val_opt) = db.get(column, &buf[..]) {
```

Comment:
That is way to complicated. Either split it up across multiple lines with `?` or use match.